sycl/include/sycl/detail/accessor_iterator.hpp

```diff
@@ -0,0 +1,439 @@
+    // iterator && N > 0 -> backwards
+    // iterator && N < 0 -> forward
+    // reverse iterator && N > 0 -> forward
+    // reverse iterator && N < 0 -> backwards
+    bool _ForwardDirection = !_IsReverse ^ (_N > 0);
+    if (_ForwardDirection)
+      __adjustForward(_N);
+    else
+      __adjustBackwards(_N);
+    return *this;
+  }
+
+  difference_type operator-(const __accessor_iterator_base &_Rhs) {
+    if (_Rhs._MLinearCurrent > _MLinearCurrent)
+      return _Rhs._MLinearCurrent - _MLinearCurrent;
+    else
+      return _MLinearCurrent - _Rhs._MLinearCurrent;
+  }
+
+  bool operator<(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return _MLinearCurrent < _Other._MLinearCurrent;
+  }
+
+  bool operator>(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return _Other < *this;
+  }
+
+  bool operator<=(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return !(*this > _Other);
+  }
+
+  bool operator>=(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return !(*this < _Other);
+  }
+
+  bool operator==(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return _MLinearCurrent == _Other._MLinearCurrent;
+  }
+
+  bool operator!=(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return !(*this == _Other);
+  }
+
+private:
+  void __increment() {
+    if (_MLinearCurrent >= _MLinearEnd)
+      return;
+
+    ++_MLinearCurrent;
+    if (_MCurrent[_Index0] < _MEnd[_Index0])
+      _MCurrent[_Index0]++;
+    if constexpr (_Dimensions > 1) {
+      if (_MCurrent[_Index0] == _MEnd[_Index0]) {
+        if (_MCurrent[_Index1] < _MEnd[_Index1]) {
+          _MCurrent[_Index1]++;
+          _MCurrent[_Index0] = _MBegin[_Index0];
+        }
+      }
+    }
+    if constexpr (_Dimensions > 2) {
+      if (_MCurrent[_Index1] == _MEnd[_Index1]) {
+        if (_MCurrent[_Index2] < _MEnd[_Index2]) {
+          _MCurrent[_Index2]++;
+          _MCurrent[_Index0] = _MBegin[_Index0];
+          _MCurrent[_Index1] = _MBegin[_Index1];
+        }
+      }
+    }
+  }
+
+  void __decrement() {
+    if (_MLinearCurrent == _LinearBegin)
+      return;
+
+    --_MLinearCurrent;
+    if (_MCurrent[_Index0] > 0)
+      _MCurrent[_Index0]--;
+    if constexpr (_Dimensions > 1) {
+      if (_MCurrent[_Index0] == 0) {
+        if (_MCurrent[_Index1] > 0) {
+          _MCurrent[_Index1]--;
+          _MCurrent[_Index0] = _MEnd[_Index0] - 1;
+        }
+      }
+    }
+    if constexpr (_Dimensions > 2) {
+      if (_MCurrent[_Index1] == 0) {
+        if (_MCurrent[_Index2] > 0) {
+          _MCurrent[_Index2]--;
+          _MCurrent[_Index0] = _MEnd[_Index0] - 1;
+          _MCurrent[_Index1] = _MEnd[_Index1] - 1;
+        }
+      }
+    }
+  }
+
+  void __adjustForward(difference_type _N) {
+    if (_MLinearCurrent + _N > _MLinearEnd)
+      _MLinearCurrent = _MLinearEnd;
+    else
+      _MLinearCurrent += _N;
+    _MCurrent = __delinearizeIndex(_MLinearCurrent);
```

Comment:
Can we have a test checking the generated LLVM IR and/or target assembly to ensure the compiler toolchain can generate it efficiently (i.e. without doing division).

Reply:
I'm not entirely sure that we can add a test which checks for optimized LLVM IR, because some optimizations are done by low-level runtimes and I doubt that such test would be reliable.

We also synced offline and there is a generic comment that we shouldn't pay the price of handling ranged accessors with we are dealing with non-ranged accessors. The latter should be as much as possible close to a plain pointer. I have an idea how it can be done, I will post an update to the implementation.

Reply:
I've rewritten the implementation, so it is now should be as cheap as possible for non-ranged accessors, see ba4557784b2781a6b750b3489e1d6cf850b4a615